File: hooks/scripts/cmd/wisdom_accumulator.go
package cmd

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/spf13/cobra"
	"solon-hooks/internal/config"
	"solon-hooks/internal/hookio"
)

const maxWisdomLines = 50

var wisdomAccumulatorCmd = &cobra.Command{
	Use:   "wisdom-accumulator",
	Short: "Extract and store learnings from completed subagents",
	RunE:  runWisdomAccumulator,
}

func runWisdomAccumulator(cmd *cobra.Command, args []string) error {
	if !config.IsHookEnabled("wisdom-accumulation") {
		os.Exit(0)
	}

	defer func() {
		if r := recover(); r != nil {
			os.Exit(0)
		}
	}()

	var input hookio.SubagentStopInput
	_ = hookio.ReadInput(&input)

	planPath := os.Getenv("SL_ACTIVE_PLAN")
	sessionID := input.SessionID
	if sessionID == "" {
		sessionID = os.Getenv("SL_SESSION_ID")
	}

	// Resolve wisdom file path
	var wisdomPath string
	if planPath != "" {
		wisdomPath = filepath.Join(planPath, ".wisdom.md")
	} else if sessionID != "" {
		wisdomPath = filepath.Join(os.TempDir(), "sl-wisdom-"+sessionID+".md")
	} else {
		os.Exit(0)
	}

	// Extract learnings from transcript tail
	learnings := extractLearnings(input.TranscriptPath, input.AgentType)
	if learnings == "" {
		os.Exit(0)
	}

	// Append to wisdom file and prune if needed
	appendWisdom(wisdomPath, learnings)
	pruneWisdom(wisdomPath, maxWisdomLines)

	return nil
}

// extractLearnings reads last 30 lines of transcript and extracts actionable items.
func extractLearnings(transcriptPath, agentType string) string {
	if transcriptPath == "" {
		return ""
	}

	f, err := os.Open(transcriptPath)
	if err != nil {
		return ""
	}
	defer f.Close()

	// Read all lines, keep last 30
	var lines []string
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024) // 1MB buffer
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if len(lines) > 30 {
		lines = lines[len(lines)-30:]
	}

	// Extract lines containing actionable keywords
	keywords := []string{
		"learned", "gotcha", "important", "note:", "warning:",
		"convention", "pattern", "decision", "discovered",
		"must", "should", "avoid", "don't", "careful",
		"works", "doesn't work", "failed", "succeeded",
	}

	var learnings []string
	for _, line := range lines {
		lower := strings.ToLower(line)
		for _, kw := range keywords {
			if strings.Contains(lower, kw) {
				trimmed := strings.TrimSpace(line)
				if len(trimmed) > 10 && len(trimmed) < 200 {
					learnings = append(learnings, trimmed)
				}
				break
			}
		}
	}

	if len(learnings) == 0 {
		return ""
	}

	// Cap at 5 learnings per subagent
	if len(learnings) > 5 {
		learnings = learnings[len(learnings)-5:]
	}

	header := fmt.Sprintf("### %s (%s)", agentType, time.Now().Format("15:04"))
	return header + "\n" + strings.Join(learnings, "\n")
}

// appendWisdom appends content to the wisdom file.
func appendWisdom(path, content string) {
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()
	fmt.Fprintf(f, "\n%s\n", content)
}

// pruneWisdom keeps only the last maxLines of the wisdom file.
func pruneWisdom(path string, maxLines int) {
	data, err := os.ReadFile(path)
	if err != nil {
		return
	}

	lines := strings.Split(string(data), "\n")
	if len(lines) <= maxLines {
		return
	}

	pruned := strings.Join(lines[len(lines)-maxLines:], "\n")
	os.WriteFile(path, []byte(pruned), 0644)
}

File: crates/solon-core/src/hooks/wisdom_accumulator.rs
use crate::hooks::{get_env, get_str, is_hook_enabled};
/// SubagentStop hook: extract learnings from transcript tail and append to wisdom file.
use anyhow::Result;

const MAX_WISDOM_LINES: usize = 50;

pub fn run() -> Result<()> {
    if !is_hook_enabled("wisdom-accumulation") {
        std::process::exit(0);
    }

    let input = match crate::hooks::read_hook_input() {
        Ok(v) => v,
        Err(_) => serde_json::Value::Null,
    };

    let plan_path = get_env("SL_ACTIVE_PLAN").unwrap_or_default();
    let session_id = {
        let s = get_str(&input, "session_id").to_string();
        if s.is_empty() {
            get_env("SL_SESSION_ID").unwrap_or_default()
        } else {
            s
        }
    };

    let wisdom_path = if !plan_path.is_empty() {
        std::path::PathBuf::from(&plan_path).join(".wisdom.md")
    } else if !session_id.is_empty() {
        std::env::temp_dir().join(format!("sl-wisdom-{}.md", session_id))
    } else {
        std::process::exit(0);
    };

    let transcript_path = get_str(&input, "transcript_path").to_string();
    let agent_type = get_str(&input, "agent_type").to_string();

    let learnings = extract_learnings(&transcript_path, &agent_type);
    if learnings.is_empty() {
        std::process::exit(0);
    }

    append_wisdom(&wisdom_path, &learnings);
    prune_wisdom(&wisdom_path, MAX_WISDOM_LINES);

    Ok(())
}

fn extract_learnings(transcript_path: &str, agent_type: &str) -> String {
    if transcript_path.is_empty() {
        return String::new();
    }

    let data = match std::fs::read_to_string(transcript_path) {
        Ok(d) => d,
        Err(_) => return String::new(),
    };

    let all_lines: Vec<&str> = data.lines().collect();
    let start = if all_lines.len() > 30 {
        all_lines.len() - 30
    } else {
        0
    };
    let lines = &all_lines[start..];

    let keywords = [
        "learned",
        "gotcha",
        "important",
        "note:",
        "warning:",
        "convention",
        "pattern",
        "decision",
        "discovered",
        "must",
        "should",
        "avoid",
        "don't",
        "careful",
        "works",
        "doesn't work",
        "failed",
        "succeeded",
    ];

    let mut learnings: Vec<String> = lines
        .iter()
        .filter_map(|line| {
            let lower = line.to_lowercase();
            if keywords.iter().any(|kw| lower.contains(kw)) {
                let trimmed = line.trim().to_string();
                if trimmed.len() > 10 && trimmed.len() < 200 {
                    return Some(trimmed);
                }
            }
            None
        })
        .collect();

    if learnings.is_empty() {
        return String::new();
    }
    if learnings.len() > 5 {
        let start = learnings.len() - 5;
        learnings = learnings[start..].to_vec();
    }

    let now = chrono::Local::now();
    let header = format!("### {} ({})", agent_type, now.format("%H:%M"));
    format!("{}\n{}", header, learnings.join("\n"))
}

fn append_wisdom(path: &std::path::Path, content: &str) {
    use std::io::Write;
    if let Ok(mut f) = std::fs::OpenOptions::new()
        .append(true)
        .create(true)
        .open(path)
    {
        let _ = writeln!(f, "\n{}", content);
    }
}

fn prune_wisdom(path: &std::path::Path, max_lines: usize) {
    let data = match std::fs::read_to_string(path) {
        Ok(d) => d,
        Err(_) => return,
    };
    let lines: Vec<&str> = data.split('\n').collect();
    if lines.len() <= max_lines {
        return;
    }
    let pruned = lines[lines.len() - max_lines..].join("\n");
    let _ = std::fs::write(path, pruned);
}
